watchtower-plugin/src/retrier.rs

```diff
@@ -226,22 +383,53 @@ impl Retrier {
                     // unsuccessfully or by manually raising such an error (like when facing a tower misbehavior).
                     log::warn!("Retry strategy gave up for {}. {}", self.tower_id, e);
 
-                    // Retrier failed and should be given up on. Avoid setting the tower status until the retrier is
-                    // deleted/dropped. This way users performing manual retry will get an error as the tower will be
-                    // temporary unreachable.
-                    // We don't need to set the tower status now. Any new appointments we receive will not be retried anyways.
-                    self.set_status(RetrierStatus::Failed);
+                    if e.is_permanent() {
+                        self.set_status(RetrierStatus::Failed);
+                    } else {
+                        log::debug!("Starting to idle");
+                        self.set_status(RetrierStatus::Idle(Instant::now()));
+                        // Clear all pending appointments so they do not waste any memory while idling
+                        self.pending_appointments.lock().unwrap().clear();
+                        self.wt_client
+                            .lock()
+                            .unwrap()
+                            .set_tower_status(self.tower_id, crate::TowerStatus::Unreachable);
+                    }
+
+                    match e {
+                        RetryError::Register(_) => {
+                            log::info!("Setting {} status as subscription error", self.tower_id);
+                            self.wt_client
+                                .lock()
+                                .unwrap()
+                                .set_tower_status(self.tower_id, TowerStatus::SubscriptionError)
+                        }
```

Comment:
Okay so let's assume for a second that `RetryError::Register` is permanent and it occurred.
So we will do:
```rust
if e.is_permanent() {
	// This will remove the retrier from the active retriers map held by `WTClient`.
	// Now users will be granted manual retry based on `TowerStatus::is_retryable`.
	self.set_status(RetrierStatus::Failed);
}
```

Now in the match:
```rust
match e
	RetryError::Register(_) => {
		log::info!("Setting {} status as subscription error", self.tower_id);
		self.wt_client
			.lock()
			.unwrap()
			// We set the tower status to subscription error which is `retryable`.
			.set_tower_status(self.tower_id, TowerStatus::SubscriptionError)
	}
	...
```

We effectively created the same manual retry race condition we had before.

The user will be allowed to retry (neither the retrier is in the map of active retriers held by `WTClient`, nor its tower is in a non-retryable state), but if a user actually issues a manual retry at this point (pre failed retrier being removed by the retry manager), the request will go through but the retry manager will ignore it anyways and remove the failed retrier in the next cycle.

Reply:
@sr-gi Does this make more sense?

This should be general over any permanent error and not just `Register` (if it even were to be permanent).

Generally:
permanent error -> failed retrier
fail retrier -> unsafe to set the tower status

Reply:
Ok, my bad, I failed to see that `Retrier::set_status` was still removing the data from `WTClient::retriers`. I thought I moved it to `Retrier::remove_if_failed`.

I think we can simply update `remove_if_failed` to `remove_if_done`, and make it remove data from `retriers` either if the retrier has failed or if the retrier has finished (status == Stopped and no pending appointments). Then simply take out the retriers removing part from `set_status`.

Otherwise, we need to carry around whether an error is permanent out of `Retrier::start` and check it by the `RetryManager`

```rust
fn set_status(&self, status: RetrierStatus) {
    *self.status.lock().unwrap() = status.clone();

    // Add active retriers to the WTClient
    if self.is_running() || self.is_idle() {
        log::debug!("Adding {} to active retriers", self.tower_id);
        self.wt_client
            .lock()
            .unwrap()
            .retriers
            .insert(self.tower_id, status);
    }
}
```

```rust
/// Removed our retrier identifier from the WTClient if the retrier is done (either stopped with no appointments or failed)
pub fn remove_if_done(&self) {
    if self.failed() | !self.should_start() {
        log::debug!("Removing {} from active retriers", self.tower_id);
        self.wt_client
            .lock()
            .unwrap()
            .retriers
            .remove(&self.tower_id);
    }
}
```

Reply:
Notice that, if done like this, the retrier won't be removed from `retriers` if it's stopped and has pending data, but in this case the retrier will be re-started in the next iteration cycle, so we may be saving one operation against `retriers`:

`sellf.stopped`
- with pending appointments -> remove, re-add in the next cycle
- without pending appointments -> remove

`self.should_start`
- true -> retrier stays in `retriers` as stopped and is updated in the next iteration (it's a hashmap)
- false -> remove

Reply:
Ok, no, nvm, this actually moves the race condition to another place of the codebase. The retrier should be removed from `retriers` when it is stopped, otherwise, the race condition will exist in the `idle->stopped->running` transition.

`pending_appointments` will be loaded when the `idle-stopped` transition happens, but this is not reflected in the `retriers` map, so `send_to_retrier` may miss some data. This should be better:

```rust
fn set_status(&self, status: RetrierStatus) {
    *self.status.lock().unwrap() = status.clone();

    // Add or remove retriers from WTClient based on the RetrierStatus
    if self.is_running() || self.is_idle() {
        log::debug!("Adding {} to active retriers", self.tower_id);
        self.wt_client
            .lock()
            .unwrap()
            .retriers
            .insert(self.tower_id, status);
    } else if self.is_stopped() {
        log::debug!("Removing retrier {} from active retriers", self.tower_id);
        self.wt_client
            .lock()
            .unwrap()
            .retriers
            .remove(&self.tower_id);
    }
}
```

```rust
/// Removed our retrier identifier from the WTClient if the retrier has failed
pub fn remove_if_failed(&self) {
    if self.failed() {
        log::debug!(
            "Removing failed retrier {} from active retriers",
            self.tower_id
        );
        self.wt_client
            .lock()
            .unwrap()
            .retriers
            .remove(&self.tower_id);
    }
}
```

Reply:
@sr-gi That would do.

Another solution I think of is identifying user-made retry requests and force execute them (in even in failed retriers). This is easy now since we have an enum for the revocation data.

Approaching it this way would relief us from thinking about whether a race condition will happen in manual retry or not, and we can safely set the tower status from any retrier/thread other than the RetryManager knowing that if a user executes a retry request on this very same tower, the RetryManager will force-execute their request.

Some (incomplete/pseudocode) example of what I mean:
```rust
loop {
    match self.unreachable_towers.try_recv() {
        Ok((tower_id, data)) => {
            match data {
                // None for user-made retry requests.
                RevocationData::None => {
                    if let Some(retrier) = self.retriers.get(&tower_id) {
                        // Indicates that the retry manager should force start that
                        // retrier in the next retry cycle.
                        retrier.force_start = true;
                        // Now load the pending appointments and extend the retrier with them.
                        retrier.pending_appointments.extend(
                            load_appointment_locators(
                                tower_id,
                                crate::AppointmentStatus::Pending,
                            ),
                        );
                    } else {
                        // Just add a new retrier
                        // ...
                    }
                }
                // Handle stale and fresh ...
                _ => unreachable!()
            }
        }
        Err(TryRecvError::Empty) => {
            // 1- Remove useless retriers (but keep force_start ones even if failed).
            // ...
            // 2- Start the shut down retriers one by one.
            for retrier in self.retriers.values() {
                if retrier.force_start {
                    retrier.set_status(RetrierStatus::Stopped);
                    retrier.force_start = false;
                }
                if retrier.should_start() {
                    self.start_retrying(retrier.clone())
                }
            }
        }
        _ => panic!(),
}
```

Reply:
One more solution is to allow or disallow manual retries based on retriers and not tower status.

Thus if there is a retrier for tower X we don't let the user retry (the retrier in any status generally - other than idle) and if there is not, the user can sent `RevocationData::None` peacefully.

Reply:
I like the idea, but I think it requires a pretty big refactor overall, doesn't it?

I would feel more comfortable merging this incremental change and considering a followup for a complete refactor of this logic based on the `RevocationData` type. If that's something you'd like to work on I'd be more than happy to review it.

Reply:
> I like the idea, but I think it requires a pretty big refactor overall, doesn't it?

> I would feel more comfortable merging this incremental change and considering a followup > for a complete refactor of this logic based on the RevocationData type.

+1

> If that's something you'd like to work on I'd be more than happy to review it.

Yeah sure :blush:

Reply:
Leaving this as unresolved, we can open an issue to keep track of it